# HG changeset patch
# User John W. Eaton <jwe@octave.org>
# Date 1251306940 14400
#      Wed Aug 26 13:15:40 2009 -0400
# Node ID d3fccc4c4b9ea4876bd96f34d615a021670dc087
# Parent  0a4eb56c56760a34c47292cc0714702f8c7b097a
use OCTAVE_CHECK_LIBRARY to check for ARPACK

diff --git a/ChangeLog b/ChangeLog
--- a/ChangeLog
+++ b/ChangeLog
@@ -1,8 +1,17 @@
+2009-08-26  John W. Eaton  <jwe@octave.org>
+
+	* aclocal.m4 (OCTAVE_CHECK_LIBRARY): Allow empty list of header files.
+	* configure.in: Use OCTAVE_CHECK_LIBRARY to check for arpack.
+	* octave-bug.in, octave-bug.cc.in: Substitute ARPACK_CPPFLAGS and
+	ARPACK_LDFLAGS.
+	* Makeconf.in (ARPACK_CPPFLAGS, ARPACK_LDFLAGS): New variables.
+	(do-subst-config-vals): Substitute them.
+
 2009-08-26  Benjamin Lindner  <lindnerb@users.sourceforge.net>
 
 	* mkoctfile.cc.in: Fix typo.
 	* octave-bug.cc.in: Correct substututions.
 
 2009-08-24  Jaroslav Hajek  <highegg@gmail.com>
 
 	* acx_blas_f77_func.m4: Fix typo.
diff --git a/Makeconf.in b/Makeconf.in
--- a/Makeconf.in
+++ b/Makeconf.in
@@ -287,17 +287,21 @@ CXSPARSE_LDFLAGS = @CXSPARSE_LDFLAGS@
 CXSPARSE_LIBS = @CXSPARSE_LIBS@
 
 UMFPACK_LIBS = @UMFPACK_LIBS@
 
 OPENGL_LIBS = @OPENGL_LIBS@
 QRUPDATE_LIBS = @QRUPDATE_LIBS@
 READLINE_LIBS = @READLINE_LIBS@
 TERM_LIBS = @TERM_LIBS@
+
+ARPACK_CPPFLAGS = @ARPACK_CPPFLAGS@
+ARPACK_LDFLAGS = @ARPACK_LDFLAGS@
 ARPACK_LIBS = @ARPACK_LIBS@
+
 DL_LIBS = @DL_LIBS@
 LIBS = @LIBS@
 
 ALL_CPPFLAGS = $(CPPFLAGS) $(HDF5_CPPFLAGS) $(Z_CPPFLAGS)
 
 ## Use sort to remove duplicates.
 SPARSE_CPPFLAGS = \
   $(sort $(AMD_CPPFLAGS) $(CAMD_CPPFLAGS) $(COLAMD_CPPFLAGS) \
@@ -551,16 +555,18 @@ echo "making $@ from $<"
   -e "s|%OCTAVE_CONF_ALL_CXXFLAGS%|\"${ALL_CXXFLAGS}\"|" \
   -e "s|%OCTAVE_CONF_ALL_FFLAGS%|\"${ALL_FFLAGS}\"|" \
   -e "s|%OCTAVE_CONF_ALL_LDFLAGS%|\"${ALL_LDFLAGS}\"|" \
   -e "s|%OCTAVE_CONF_AMD_CPPFLAGS%|\"${AMD_CPPFLAGS}\"|" \
   -e "s|%OCTAVE_CONF_AMD_LDFLAGS%|\"${AMD_LDFLAGS}\"|" \
   -e "s|%OCTAVE_CONF_AMD_LIBS%|\"${AMD_LIBS}\"|" \
   -e "s|%OCTAVE_CONF_AR%|\"${AR}\"|" \
   -e "s|%OCTAVE_CONF_ARFLAGS%|\"${ARFLAGS}\"|" \
+  -e "s|%OCTAVE_CONF_ARPACK_CPPFLAGS%|\"${ARPACK_CPPFLAGS}\"|" \
+  -e "s|%OCTAVE_CONF_ARPACK_LDFLAGS%|\"${ARPACK_LDFLAGS}\"|" \
   -e "s|%OCTAVE_CONF_ARPACK_LIBS%|\"${ARPACK_LIBS}\"|" \
   -e "s|%OCTAVE_CONF_BLAS_LIBS%|\"${BLAS_LIBS}\"|" \
   -e "s|%OCTAVE_CONF_CAMD_CPPFLAGS%|\"${CAMD_CPPFLAGS}\"|" \
   -e "s|%OCTAVE_CONF_CAMD_LDFLAGS%|\"${CAMD_LDFLAGS}\"|" \
   -e "s|%OCTAVE_CONF_CAMD_LIBS%|\"${CAMD_LIBS}\"|" \
   -e "s|%OCTAVE_CONF_CANONICAL_HOST_TYPE%|\"${canonical_host_type}\"|" \
   -e "s|%OCTAVE_CONF_CARBON_LIBS%|\"${CARBON_LIBS}\"|" \
   -e "s|%OCTAVE_CONF_CC%|\"${CC}\"|" \
diff --git a/aclocal.m4 b/aclocal.m4
--- a/aclocal.m4
+++ b/aclocal.m4
@@ -464,17 +464,20 @@ AC_DEFUN(OCTAVE_CHECK_LIBRARY, [
   esac
 
   [TEXINFO_]m4_toupper([$1])=
   warn_$1="$3"
   if test -n "$m4_toupper([$1])_LIBS"; then
     octave_check_library_save_CPPFLAGS="$CPPFLAGS"
     CPPFLAGS="$m4_toupper([$1])_CPPFLAGS $CPPFLAGS"
     m4_ifnblank([$6], [AC_LANG_PUSH($6)])
-    AC_CHECK_HEADERS($4, [
+    octave_$1_check_for_lib=false
+    m4_ifblank([$4], [octave_$1_check_for_lib=true],
+               [AC_CHECK_HEADERS($4, [octave_$1_check_for_lib=true; break])])
+    if $octave_$1_check_for_lib; then
       octave_check_library_save_LDFLAGS="$LDFLAGS"
       LDFLAGS="$m4_toupper([$1])_LDFLAGS $LDFLAGS"
       octave_check_library_save_LIBS="$LIBS"
       LIBS="$m4_toupper([$1])_LIBS $LIBS"
       octave_$1_ok=no
       AC_MSG_CHECKING([for $5 in $m4_toupper([$1])_LIBS])
       AC_LINK_IFELSE([AC_LANG_CALL([], [$5])],
 	[octave_$1_ok=yes])
@@ -482,17 +485,18 @@ AC_DEFUN(OCTAVE_CHECK_LIBRARY, [
       if test $octave_$1_ok = yes; then
 	m4_ifblank([$8], [
 	  warn_$1=
 	  AC_DEFINE([HAVE_]m4_toupper([$1]), 1,
             [Define if $2 is available.])
 	  [TEXINFO_]m4_toupper([$1])="@set [HAVE_]m4_toupper([$1])"], [$8])
       fi
       LIBS="$octave_check_library_save_LIBS"
-      LDFLAGS="$octave_check_library_save_LDFLAGS"])
+      LDFLAGS="$octave_check_library_save_LDFLAGS"
+    fi
     m4_ifnblank([$6], [AC_LANG_POP($6)])
     CPPFLAGS="$octave_check_library_save_CPPFLAGS"
   fi
   AC_SUBST(m4_toupper([$1])_LIBS)
   AC_SUBST([TEXINFO_]m4_toupper([$1]))
   if test -n "$warn_$1"; then
     AC_MSG_WARN($warn_$1)
     m4_toupper([$1])_LIBS=
diff --git a/configure.in b/configure.in
--- a/configure.in
+++ b/configure.in
@@ -931,36 +931,23 @@ if test "$with_umfpack" = yes && test "$
     break])
 fi
 if test -n "$warn_umfpack"; then
   AC_MSG_WARN($warn_umfpack)
 fi
 
 AC_SUBST(TEXINFO_UMFPACK)
 
-ARPACK_LIBS=
-AC_SUBST(ARPACK_LIBS)
-
-AC_ARG_WITH(arpack,
-  [AS_HELP_STRING([--without-arpack],
-     [don't use ARPACK library, disable some sparse functionality])],
-  with_arpack=$withval, with_arpack=yes)
-
-warn_arpack="arpack not found. This will result in a lack of the eigs function."
-if test "$with_arpack" = yes; then
-  with_arpack=no
-  AC_CHECK_LIB(arpack, F77_FUNC(dseupd,DSEUPD), [ARPACK_LIBS="-larpack"; with_arpack=yes], , $LAPACK_LIBS $FLIBS)
-  if test "$with_arpack" = yes; then
-    AC_DEFINE(HAVE_ARPACK, 1, [Define if the ARPACK library is used.])
-    warn_arpack=
-  fi
-fi
-if test -n "$warn_arpack"; then
-  AC_MSG_WARN($warn_arpack)
-fi
+save_LIBS="$LIBS"
+LIBS="$LAPACK_LIBS $BLAS_LIBS $FLIBS $LIBS"
+OCTAVE_CHECK_LIBRARY(arpack, ARPACK,
+  [arpack not found.  The eigs function will be disabled.],
+  [],
+  [F77_FUNC(dseupd,DSEUPD)],
+  [], [don't use the ARPACK library, disable eigs function])
 
 ### Handle shared library options.
 
 ### Enable creation of static libraries.
 
 AC_ARG_ENABLE(static,
   [AS_HELP_STRING([--enable-static], [create static libraries])],
   [if test "$enableval" = no; then STATIC_LIBS=false;
diff --git a/octave-bug.cc.in b/octave-bug.cc.in
--- a/octave-bug.cc.in
+++ b/octave-bug.cc.in
@@ -99,16 +99,18 @@ main (int argc, char **argv)
   vars["RLD_FLAG"] = %OCTAVE_CONF_RLD_FLAG%;
   vars["LIBS"] = %OCTAVE_CONF_LIBS%;
   vars["LEXLIB"] = %OCTAVE_CONF_LEXLIB%;
   vars["LIBGLOB"] = %OCTAVE_CONF_LIBGLOB%;
 
   vars["AMD_CPPFLAGS"] = %OCTAVE_CONF_AMD_CPPFLAGS%;
   vars["AMD_LDFLAGS"] = %OCTAVE_CONF_AMD_LDFLAGS%;
   vars["AMD_LIBS"] = %OCTAVE_CONF_AMD_LIBS%;
+  vars["ARPACK_CPPFLAGS"] = %OCTAVE_CONF_ARPACK_CPPFLAGS%;
+  vars["ARPACK_LDFLAGS"] = %OCTAVE_CONF_ARPACK_LDFLAGS%;
   vars["ARPACK_LIBS"] = %OCTAVE_CONF_ARPACK_LIBS%;
   vars["BLAS_LIBS"] = %OCTAVE_CONF_BLAS_LIBS%;
   vars["CAMD_CPPFLAGS"] = %OCTAVE_CONF_CAMD_CPPFLAGS%;
   vars["CAMD_LDFLAGS"] = %OCTAVE_CONF_CAMD_LDFLAGS%;
   vars["CAMD_LIBS"] = %OCTAVE_CONF_CAMD_LIBS%;
   vars["CARBON_LIBS"] = %OCTAVE_CONF_CARBON_LIBS%;
   vars["CCOLAMD_CPPFLAGS"] = %OCTAVE_CONF_CCOLAMD_CPPFLAGS%;
   vars["CCOLAMD_LDFLAGS"] = %OCTAVE_CONF_CCOLAMD_LDFLAGS%;
diff --git a/octave-bug.in b/octave-bug.in
--- a/octave-bug.in
+++ b/octave-bug.in
@@ -51,16 +51,18 @@ LIBFLAGS=%OCTAVE_CONF_LIBFLAGS%
 RLD_FLAG=%OCTAVE_CONF_RLD_FLAG%
 LEXLIB=%OCTAVE_CONF_LEXLIB%
 LIBGLOB=%OCTAVE_CONF_LIBGLOB%
 LIBS=%OCTAVE_CONF_LIBS%
 
 AMD_CPPFLAGS=%AMD_CPPFLAGS%
 AMD_LDFLAGS=%AMD_LDFLAGS%
 AMD_LIBS=%AMD_LIBS%
+ARPACK_CPPFLAGS=%ARPACK_CPPFLAGS%
+ARPACK_LDFLAGS=%ARPACK_LDFLAGS%
 ARPACK_LIBS=%ARPACK_LIBS%
 BLAS_LIBS=%BLAS_LIBS%
 CAMD_CPPFLAGS=%CAMD_CPPFLAGS%
 CAMD_LDFLAGS=%CAMD_LDFLAGS%
 CAMD_LIBS=%CAMD_LIBS%
 CARBON_LIBS=%CARBON_LIBS%
 CCOLAMD_CPPFLAGS=%CCOLAMD_CPPFLAGS%
 CCOLAMD_LDFLAGS=%CCOLAMD_LDFLAGS%
diff --git a/src/ChangeLog b/src/ChangeLog
--- a/src/ChangeLog
+++ b/src/ChangeLog
@@ -1,8 +1,13 @@
+2009-08-26  John W. Eaton  <jwe@octave.org>
+
+	* oct-conf.h.in: New defines for ARPACK_CPPFLAGS and ARPACK_LDFLAGS.
+	* toplev.cc (octave_config_info): Include them in the list.
+
 2009-08-26  John W. Eaton  <jwe@octave.org>
 
 	* Makefile.in (OCTAVE_LIBS, OCTINTERP_LINK_DEPS):
 	Include BLAS_LIBS in the list.
 
 2009-08-26  John W. Eaton  <jwe@octave.org>
 
 	* Makefile.in (OCTAVE_LIBS): Include FLIBS in the list.
diff --git a/src/oct-conf.h.in b/src/oct-conf.h.in
--- a/src/oct-conf.h.in
+++ b/src/oct-conf.h.in
@@ -56,16 +56,24 @@ along with Octave; see the file COPYING.
 #ifndef OCTAVE_CONF_ARFLAGS
 #define OCTAVE_CONF_ARFLAGS %OCTAVE_CONF_ARFLAGS%
 #endif
 
 #ifndef OCTAVE_CONF_AR
 #define OCTAVE_CONF_AR %OCTAVE_CONF_AR%
 #endif
 
+#ifndef OCTAVE_CONF_ARPACK_CPPFLAGS
+#define OCTAVE_CONF_ARPACK_CPPFLAGS %OCTAVE_CONF_ARPACK_CPPFLAGS%
+#endif
+
+#ifndef OCTAVE_CONF_ARPACK_LDFLAGS
+#define OCTAVE_CONF_ARPACK_LDFLAGS %OCTAVE_CONF_ARPACK_LDFLAGS%
+#endif
+
 #ifndef OCTAVE_CONF_ARPACK_LIBS
 #define OCTAVE_CONF_ARPACK_LIBS %OCTAVE_CONF_ARPACK_LIBS%
 #endif
 
 #ifndef OCTAVE_CONF_BLAS_LIBS
 #define OCTAVE_CONF_BLAS_LIBS %OCTAVE_CONF_BLAS_LIBS%
 #endif
 
diff --git a/src/toplev.cc b/src/toplev.cc
--- a/src/toplev.cc
+++ b/src/toplev.cc
@@ -1195,16 +1195,18 @@ specified option.\n\
       { false, "ALL_CXXFLAGS", OCTAVE_CONF_ALL_CXXFLAGS },
       { false, "ALL_FFLAGS", OCTAVE_CONF_ALL_FFLAGS },
       { false, "ALL_LDFLAGS", OCTAVE_CONF_ALL_LDFLAGS },
       { false, "AMD_CPPFLAGS", OCTAVE_CONF_AMD_CPPFLAGS },
       { false, "AMD_LDFLAGS", OCTAVE_CONF_AMD_LDFLAGS },
       { false, "AMD_LIBS", OCTAVE_CONF_AMD_LIBS },
       { false, "AR", OCTAVE_CONF_AR },
       { false, "ARFLAGS", OCTAVE_CONF_ARFLAGS },
+      { false, "ARPACK_CPPFLAGS", OCTAVE_CONF_ARPACK_CPPFLAGS },
+      { false, "ARPACK_LDFLAGS", OCTAVE_CONF_ARPACK_LDFLAGS },
       { false, "ARPACK_LIBS", OCTAVE_CONF_ARPACK_LIBS },
       { false, "BLAS_LIBS", OCTAVE_CONF_BLAS_LIBS },
       { false, "CARBON_LIBS", OCTAVE_CONF_CARBON_LIBS },
       { false, "CAMD_CPPFLAGS", OCTAVE_CONF_CAMD_CPPFLAGS },
       { false, "CAMD_LDFLAGS", OCTAVE_CONF_CAMD_LDFLAGS },
       { false, "CAMD_LIBS", OCTAVE_CONF_CAMD_LIBS },
       { false, "CC", OCTAVE_CONF_CC },
       { false, "CC_VERSION", OCTAVE_CONF_CC_VERSION },
